hal/src/thumbv7em/clock/v2.rs

```diff
@@ -1,98 +1,859 @@
+//! #     pac::Peripherals,
+//! #     time::U32Ext,
+//! # };
+//! # let mut pac = Peripherals::take().unwrap();
+//! # let (buses, clocks, tokens) = clock_system_at_reset(
+//! #     pac.OSCCTRL,
+//! #     pac.OSC32KCTRL,
+//! #     pac.GCLK,
+//! #     pac.MCLK,
+//! #     &mut pac.NVMCTRL,
+//! # );
+//! let pins = Pins::new(pac.PORT);
+//! let xosc0 = Xosc::from_crystal(
+//!     tokens.xosc0,
+//!     pins.pa14,
+//!     pins.pa15,
+//!     8.mhz(),
+//! ).enable();
+//! ```
+//!
+//! Next, we want to use a DPLL to multiply the 8 MHz crystal clock up to 100
+//! MHz. Once again, we need to decide between two instances of a clock, because
+//! each chip has two [`Dpll`]s. This time, however, our decision between
+//! [`Dpll0`] and [`Dpll1`] is arbitrary.
+//!
+//! Also note that, like before, `Dpll0<I>` and `Dpll1<I>` are aliases for
+//! `Dpll<Dpll0Id, I>` and `Dpll<Dpll1Id, I>`. [`Dpll0Id`] and [`Dpll1Id`]
+//! represent the *identity* of the respective DPLL, while `I` represents the
+//! [`Id` type](self#id-types) for the [`Source`] driving the DPLL. In this
+//! particular case, we aim to create an instance of `Dpll0<Xosc0Id>`.
 //!
-//! This module provides the set of abstractions allowing a user to manipulate
-//! an ATSAMD's clocking system in a safe and expressive manner.
+//! Only certain clocks can drive the DPLL, so `I` is constrained by the
+//! [`DpllSourceId`] trait. Specifically, only the [`Xosc0Id`], [`Xosc1Id`],
+//! [`Xosc32kId`] and [`GclkId`] types implement this trait.
 //!
-//! Foundation of an API is a [`retrieve_clocks`] function that returns a tuple
-//! of instantiated clocking abstractions that are enabled on reset.
+//! As before, we access the [`Tokens`] struct and use the corresponding
+//! [`DpllToken`] when creating an instance of `Dpll`. However, unlike before,
+//! we are creating a new clock-tree relationship that must be tracked by the
+//! type system. Because DPLL0 will now consume XOSC0, we must [`Increment`]
+//! the [`Enabled`] [`Counter`] for [`EnabledXosc0`].
 //!
-//! To reconfigure a clocking tree, user can instantiate additional clocking
-//! components or reconfigure existing ones and connect them. Type-safe nature
-//! of the API prevents user from building an application with an unsound clock
-//! tree setup.
+//! Thus, to create an instance of `Dpll0<XoscId0>`, we must provide the
+//! `EnabledXosc0`, so that its `U0` type parameter can be incremented to `U1`.
+//! The `Dpll::from_source` method takes ownership of the `EnabledXosc0` and
+//! returns it with this modified type parameter.
+//!
+//! This is the essence of clock safety in this module. Once the `Counter` has
+//! been incremeneted to `U1`, the `EnabledXosc0` can no longer be modified or
+//! disabled. All further code can guarantee this invariant is upheld. To modify
+//! the `EnabledXosc0`, we would first have to use `Dpll::free_source` to
+//! disable the DPLL and [`Decrement`] the `Counter` back to `U0`.
```

Comment:
> (..) and `Counter` would be `Decrement`ed back to `U0`. 

I'm referring to previous comment of mine.

Although we also discussed removal of `Counter` trait so probably that will get totally reworded